File: broca-web/frontend/src/api/crew.ts
import request from '@/utils/request'

/**
 * 编排器类型
 */
export const OrchestratorType = {
  PIPELINE: 'pipeline',
  SUPERVISOR_WORKER: 'supervisor-worker',
  ROUND_TABLE: 'round-table',
  BROADCAST: 'broadcast',
  CONSENSUS: 'consensus',
  COMPOSITE: 'composite',
} as const

export type OrchestratorType = (typeof OrchestratorType)[keyof typeof OrchestratorType]

/**
 * 执行状态
 */
export const ExecutionStatus = {
  PENDING: 'pending',
  RUNNING: 'running',
  COMPLETED: 'completed',
  FAILED: 'failed',
  ABORTED: 'aborted',
} as const

export type ExecutionStatus = (typeof ExecutionStatus)[keyof typeof ExecutionStatus]

/**
 * Agent 角色配置
 */
export interface AgentRoleConfig {
  role: string
  name: string
  config: string
  use_history?: boolean        // 是否保留历史消息（默认 false）
  extras?: Record<string, any>
}

/**
 * 编排器配置
 */
export interface OrchestratorConfig {
  type: OrchestratorType
  max_rounds?: number
  strategy?: string
  threshold?: number
  weights?: Record<string, number>
}

/**
 * Crew 配置
 */
export interface CrewConfig {
  name: string
  description: string
  orchestrator: OrchestratorConfig
  agents: AgentRoleConfig[]
  blackboard?: {
    initial_entries?: Array<{ key: string; value: any }>
  }
  sub_crews?: Array<{
    name: string
    orchestrator: OrchestratorConfig
    steps?: Array<{ agent: string; task: string }>
  }>
}

/**
 * 阶段结果
 */
export interface PhaseResult {
  name: string
  status: string
  agents: string[]
  output?: Record<string, any>
  error?: string
  started_at?: string
  completed_at?: string
}

/**
 * 编排执行记录
 */
export interface CrewExecution {
  execution_id: string
  session_id: string
  crew_name: string
  description: string
  orchestrator_type: OrchestratorType
  agent_count: number
  status: ExecutionStatus
  error?: string
  result?: Record<string, any>
  phases?: PhaseResult[]
  phases_total?: number
  progress?: number
  created_at: string
  completed_at?: string
}

/**
 * 提交编排请求
 */
export interface CrewSubmitRequest {
  yaml_content?: string
  yaml_path?: string
  session_id: string
}

/**
 * 校验编排请求
 */
export interface CrewValidateRequest {
  yaml_content?: string
  yaml_path?: string
}

/**
 * 校验结果
 */
export interface CrewValidateResult {
  valid: boolean
  errors: string[]
  error_count: number
}

/**
 * 编排列表响应
 */
export interface CrewListResponse {
  executions: CrewExecution[]
  total: number
}

// ============================================================================
// Workspace crew_configs 相关类型
// ============================================================================

/**
 * 编排配置文件摘要信息
 */
export interface CrewConfigFile {
  filename: string
  path: string
  name: string
  description: string
  orchestrator_type: string | null
  agent_count: number
  agent_names: string[]
  modified_time: number
  parse_error?: string
}

/**
 * 编排配置文件详情（含完整内容）
 */
export interface CrewConfigDetail {
  filename: string
  path: string
  content: string
  summary: {
    name?: string
    description?: string
    orchestrator_type?: string | null
    agent_count?: number
    agent_names?: string[]
    parse_error?: string
  }
  modified_time: number
}

/**
 * Crew API
 */
export const crewApi = {
  /**
   * 提交编排执行
   */
  async submit(data: CrewSubmitRequest): Promise<CrewExecution> {
    return request.post('/crews', data)
  },

  /**
   * 校验编排配置
   */
  async validate(data: CrewValidateRequest): Promise<CrewValidateResult> {
    return request.post('/crews/validate', data)
  },

  /**
   * 获取编排执行列表
   */
  async list(params?: {
    session_id?: string
    status?: string
  }): Promise<CrewListResponse> {
    return request.get('/crews', { params })
  },

  /**
   * 获取编排执行详情
   */
  async getDetail(executionId: string): Promise<CrewExecution> {
    return request.get(`/crews/${executionId}`)
  },

  /**
   * 中止编排执行
   */
  async abort(executionId: string): Promise<{ execution_id: string }> {
    return request.post(`/crews/${executionId}/abort`)
  },

  /**
   * 删除编排执行记录
   */
  async delete(executionId: string): Promise<{ execution_id: string }> {
    return request.delete(`/crews/${executionId}`)
  },

  // ==========================================================================
  // Workspace crew_configs 操作
  // ==========================================================================

  /**
   * 列出 workspace 下 crew_configs 目录中已有的编排配置文件
   */
  async listConfigs(workspace: string): Promise<{ configs: CrewConfigFile[]; total: number }> {
    return request.get('/crews/configs', { params: { workspace } })
  },

  /**
   * 获取 workspace crew_configs 目录下指定配置文件的内容
   */
  async getConfigDetail(filename: string, workspace: string): Promise<CrewConfigDetail> {
    return request.get(`/crews/configs/${encodeURIComponent(filename)}`, {
      params: { workspace },
    })
  },

  /**
   * 保存/更新 workspace crew_configs 目录下的配置文件
   */
  async saveConfig(filename: string, workspace: string, content: string): Promise<CrewConfigFile> {
    return request.put(`/crews/configs/${encodeURIComponent(filename)}`, {
      workspace,
      filename,
      content,
    })
  },
}

export default crewApi

File: broca-tui/broca_tui/api/crew.py
"""
Crew Execution API client.

Provides methods for submitting, listing, monitoring, and aborting crew executions.
"""

from typing import Any, Dict, Optional

from broca_tui.api.client import APIClient


class CrewAPI:
    """Crew execution API client."""

    def __init__(self, client: Optional[APIClient] = None):
        """Initialize crew API.

        Args:
            client: APIClient instance. Creates a new one if not provided.
        """
        self._client = client or APIClient()

    async def list_executions(
        self,
        session_id: Optional[str] = None,
        status: Optional[str] = None,
    ) -> Dict[str, Any]:
        """List crew executions.

        Args:
            session_id: Optional session ID filter
            status: Optional status filter (pending, running, completed, failed, aborted)

        Returns:
            Dict with 'executions' list and 'total' count.
        """
        params: Dict[str, Any] = {}
        if session_id:
            params["session_id"] = session_id
        if status:
            params["status"] = status
        return await self._client.get("/crews", params=params)

    async def get_execution_detail(self, execution_id: str) -> Dict[str, Any]:
        """Get execution detail by ID."""
        return await self._client.get(f"/crews/{execution_id}")

    async def submit_execution(
        self,
        session_id: str,
        yaml_path: Optional[str] = None,
        yaml_content: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Submit a new crew execution.

        Args:
            session_id: Session ID
            yaml_path: Path to the YAML config file
            yaml_content: Inline YAML content (alternative to yaml_path)

        Returns:
            Created CrewExecution dict.
        """
        data: Dict[str, Any] = {"session_id": session_id}
        if yaml_path:
            data["yaml_path"] = yaml_path
        if yaml_content:
            data["yaml_content"] = yaml_content
        return await self._client.post("/crews", data=data)

    async def abort_execution(self, execution_id: str) -> Dict[str, Any]:
        """Abort a running execution."""
        return await self._client.post(f"/crews/{execution_id}/abort")

    async def delete_execution(self, execution_id: str) -> Dict[str, Any]:
        """Delete an execution record."""
        return await self._client.delete(f"/crews/{execution_id}")

    async def validate_config(
        self,
        yaml_path: Optional[str] = None,
        yaml_content: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Validate a crew configuration.

        Returns:
            Dict with 'valid' bool and 'errors' list.
        """
        data: Dict[str, Any] = {}
        if yaml_path:
            data["yaml_path"] = yaml_path
        if yaml_content:
            data["yaml_content"] = yaml_content
        return await self._client.post("/crews/validate", data=data)

    # ==================== Workspace Config Files ====================

    async def list_config_files(self, workspace: str) -> Dict[str, Any]:
        """List crew config files in a workspace.

        Args:
            workspace: Workspace path

        Returns:
            Dict with 'configs' list and 'total' count.
        """
        return await self._client.get(
            "/crews/configs", params={"workspace": workspace}
        )

    async def get_config_detail(
        self, filename: str, workspace: str
    ) -> Dict[str, Any]:
        """Get detail of a config file."""
        import urllib.parse

        encoded_name = urllib.parse.quote(filename, safe="")
        return await self._client.get(
            f"/crews/configs/{encoded_name}", params={"workspace": workspace}
        )

    async def save_config(
        self, filename: str, workspace: str, content: str
    ) -> Dict[str, Any]:
        """Save/update a config file."""
        import urllib.parse

        encoded_name = urllib.parse.quote(filename, safe="")
        return await self._client.put(
            f"/crews/configs/{encoded_name}",
            data={"workspace": workspace, "filename": filename, "content": content},
        )

    async def close(self):
        """Close the underlying HTTP client."""
        await self._client.close()
